presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java

```diff
@@ -0,0 +1,234 @@
+import io.prestosql.spi.type.Type;
+import io.prestosql.spi.type.TypeUtils;
+import io.prestosql.sql.planner.plan.PlanNodeId;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+import static com.google.common.base.Verify.verify;
+import static io.prestosql.SystemSessionProperties.getDynamicFilteringMaxPerDriverByteSize;
+import static io.prestosql.SystemSessionProperties.getDynamicFilteringMaxPerDriverValueCount;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toSet;
+
+/**
+ * This operator acts as a simple "pass-through" pipe, while saving its input pages.
+ * The collected pages' value are used for creating a run-time filtering constraint (for probe-side table scan in an inner join).
+ * We support only small build-side pages (which should be the case when using "broadcast" join).
+ */
+public class DynamicFilterSourceOperator
+        implements Operator
+{
+    public static class Channel
+    {
+        private final String filterId;
+        private final Type type;
+        private final int index;
+
+        public Channel(String filterId, Type type, int index)
+        {
+            this.filterId = filterId;
+            this.type = type;
+            this.index = index;
+        }
+    }
+
+    public static class DynamicFilterSourceOperatorFactory
+            implements OperatorFactory
+    {
+        private final int operatorId;
+        private final PlanNodeId planNodeId;
+        private final Consumer<TupleDomain<String>> dynamicPredicateConsumer;
+        private final List<Channel> channels;
+
+        public DynamicFilterSourceOperatorFactory(int operatorId, PlanNodeId planNodeId, Consumer<TupleDomain<String>> dynamicPredicateConsumer, List<Channel> channels)
+        {
+            this.operatorId = operatorId;
+            this.planNodeId = requireNonNull(planNodeId, "planNodeId is null");
+            this.dynamicPredicateConsumer = requireNonNull(dynamicPredicateConsumer, "dynamicPredicateConsumer is null");
+            this.channels = requireNonNull(channels, "channels is null");
+            verify(channels.stream().map(channel -> channel.filterId).collect(toSet()).size() == channels.size(),
+                    "duplicate dynamic filters are not allowed");
+            verify(channels.stream().map(channel -> channel.index).collect(toSet()).size() == channels.size(),
+                    "duplicate channel indices are not allowed");
+        }
+
+        @Override
+        public Operator createOperator(DriverContext driverContext)
+        {
+            return new DynamicFilterSourceOperator(
+                    driverContext.addOperatorContext(operatorId, planNodeId, DynamicFilterSourceOperator.class.getSimpleName()),
+                    dynamicPredicateConsumer,
+                    channels,
+                    planNodeId);
+        }
+
+        @Override
+        public void noMoreOperators()
+        {
```

Comment:
nit: set `closed = true`. See how other operators implement this method

Reply:
Added (together with a `checkState`)